docs/principles-for-package-repository-security.md

```diff
@@ -0,0 +1,105 @@
+
+Does the package repository ...
+
+... have user accounts?
+- Yes: then you have to manage things like authentication and account recovery. For example: PyPI
+- No: For example: {index, proxy, sum}.golang.org
+
+... accept built packages, do builds on behalf of users, or only host source code?
+- Accept built packages: then you have to provide a way to link packages back to source code and build instructions. For example: npm
+- Builds on behalf of users: For example, Homebrew
+- Only host source code: For example: {index, proxy, sum}.golang.org
+
+## Security Capabilities of Package Repositories
+
+We define 4 levels of security maturity:
+- Level 0 is defined as having very little security maturity.
+- Level 1 is defined as having basic security maturity. Package repositories support basic security features like multi-factor authentication, allow security researchers to report vulnerabilities. All package management ecosystems should be working towards this level.
+- Level 2 is defined as having moderate security.
+- Level 3 is defined as having advanced security. These items are more aspirational, especially for smaller package management ecosystems.
+
+### Authentication
+
+Applies to: package repositories that have user accounts.
+
+- Level 1
+  - The package repository requires users to verify their email address. This ensures that if a user loses access to their account, they are able to recover it.
+  - The package repository documents their account recovery policy.
+  - The package repository supports multi-factor authentication (MFA) via (at minimum) TOTP.
+  - The package repository notifies maintainers via email for critical account security changes, such as password changes or disabling multi-factor authentication. This helps users detect if their account has been compromised.
+  - The package repository implements account security measures like brute force prevention (even with 2FA attempts)
+- Level 2
+  - To prevent domain resurrection for account takeover via the recovery process, the package repository detects abandoned email domains. This may look like doing a WHOIS lookup on all registered email domains, and locking accounts for which the domain is abandoned.
+  - The package repository supports MFA via phishing-resistant means such as WebAuthn
+  - The package repository requires MFA for packages deemed critical (e.g. top 1% of packages by downloads).
+  - The package repository integrates with public leaked credential databases such as Have I Been Pwned to detect when users are using leaked credentials. If a user has a leaked credential, the package repository prompts them to change it upon login.
+- Level 3
+  - The package repository supports passwordless authentication, such as passkeys.
+  - The package repository requires MFA for all maintainers.
```

Comment:
I think blanket requirements introduce friction and may be onerous for casual/new developers that want to publish hobby projects and such.

Instead individual packages should be labeled based on their level so that consumers who require higher levels can restrict them without burdening those who do not require this.

Generally the commercial/big corporate/government users shouldn't shape registries solely to their needs.
Being able to filter the content based on the criteria they're interested in should provide the same outcome.

Reply:
Authenticator apps are “proper” just like hardware keys, so i don’t think this is much of a consideration.

Reply:
Authenticator apps may sit on the same machine as the password manager or on devices that are generally more compromisable. So they provide a lower security level than dedicated hardware tokens.

AIUI webauthn also supports attestation, which means an authentication provider may choose to lock MFA to a set of trusted (and usually more expensive) vendors.

At $work we have to use physical tokens to comply with the security standards of some clients. I forsee such increasingly strict requirements will trickle down to what is deemed critical software infrastructure at some point.

Making "compliance levels" a per-package/dependency-tree instead of per-registry thing is a rather simple change that prevents an increasing security ratchet raising the barrier of entry for hobbyists and less affluent developers.

Reply:
Yes, the intent here is that MFA could either be hardware-based or TOTP. What we've seen first-hand at [GitHub's MFA rollout](https://github.blog/2023-03-09-raising-the-bar-for-software-security-github-2fa-begins-march-13/) is that this has substantially reduced account takeover while also providing global access.

Reply:
The intent is not stated here and I am concerned about this wording contributing to increasingly strict future developments.

You're using the word **requires** here. Today it's perhaps only software TOTP. Tomorrow it'll be a select set of expensive, proprietary hardware tokens. In a decade you can only get those tokens by registering with your national ID, perhaps indirectly by them being tied to a TPM in your phone.
We have seen this with telephony and SIMs. With phone booths it used to be anonymous. Then anyone could buy a SIM with cash, that got you a persistent globally trackable ID that was not directly tied to your name. Now in many countries you can only buy SIMs with an ID. All in the name of fighting crime, terrorism, money laundering or whatever.

With this recommendation and ongoing "critical software infrastructure" efforts in the EU I see this as a small step in that direction.

A step that is easy to mitigate by not making MFA a platform-wide requirement and instead something that individual projects or organizations can opt into. If a project wants to cater to high-assurance clients then they can require MFA from all their maintainers (similar to signed git commits for example) and proudly display their security posture so that their clients can trust the project.

Why would $GOV care that $HOST requires MFA for _all_ users when $GOV only wants to use project $X from $HOST? This requirement just is overly broad. It impacts those who do not share the same values as $GOV.

Reply:
Basically, the most-strict needs of some high-assurance projects shouldn't drive the baseline policy of an entire platform.

Reply:
```suggestion
  - The package repository allows individual packages to require MFA for all maintainers.
```

Reply:
> Publishing the status is something that could be an attack vector itself

Can you explain further?

Reply:
If i know which packages have MFA and which don’t, i know which ones to attack - the ones that don’t.

Reply:
That does not make it an attack vector.

Reply:
And yet, github and npm’s security teams believe it does, which is why they haven’t and likely wont reveal this info on a per-package basis.

Reply:
Do you have a source to back that up? Specifically that they consider any policy other than everyone-must-use-MFA an _attack vector_?

And at NPM mandatory MFA appears to be [opt-in at an organization level](https://docs.npmjs.com/requiring-two-factor-authentication-in-your-organization), exactly what I am suggesting.

Reply:
Both services have already announced they’re migrating to 100% mandatory MFA.

Reply:
Which still does not substantiate your claim.

Reply:
https://github.com/npm/feedback/discussions/588#discussioncomment-2932982

Reply:
That's only talking about "risk to attack", which is not the same as "attack vector". I asked you repeatedly to substantiate your claim that this constitutes an attack vector. Information usable to triaging promising targets does not constitute an attack vector.

Anyway, I consider their risk-based argument lacking nuance/detailed consideration.
For one if the non-mfa population is somewhat large then this information does not steer an attacker towards any particular target. Analogously: Just because a bank in a city block has a safe it does not put any particular private home in the environment at risk of becoming a preferred target for burglars.
It also doesn't consider second-order effects such as pushing people towards using git-based dependencies (which are an option or even the norm in some package managers) if publishing procedures become too onerous for casual users and whether that would have a net-negative effect on security.
There could also be other approaches to mitigating the concern such as only publishing that information if several additional, optional security aspects are also fulfilled. This would dilute the accuracy of the signal.
Yet another option is simply ignoring this concern because the risk is miniscule. Knowing that some niche repos don't use MFA might not be all that useful for an attacker that's looking to compromise a high-value repo.

Reply:
I’m not interested in this argument devolving into pedantry; the fact remains that it is simply not viable to require package or org specific MFA, github and npm will be requiring it wholesale imminently, and this requirement is almost certain to remain as-is in this document.

Reply:
> I’m not interested in this argument devolving into pedantry

The difference between an attack vector (which is something that needs to be fixed) and disclosure about security practices (which always is a tradeoff, potentially a tiny one) is significant in my opinion, not pedantry.

I had to ask 3 times for you to even reply to this aspect only to dismiss it as pedantry. I would expect more precise discourse when it comes to designing security documentation for global software infrastructure.

> the fact remains that it is simply not viable to require package or org specific MFA

Do you mean the requirement or publishing that information? The requirement itself is clearly _viable_ and has been for years because that's what NPM is doing right now without the world going up in fire.

There might be reasons to change the policy, but they are likely subtle/marginal ones, that are not about fundamental viability of a policy but about tradeoffs between different viable policies.
Just because github chooses a particular approach does not mean it's the only possible approach.

> and this requirement is almost certain to remain as-is in this document.

Well there was a request for feedback, which is why I'm commenting.

Anyway, I will again reiterate my suggestion that package registries should list the "compliance level" of individual packages (or even individual artifacts) instead of chosing "levels" as a baseline for the entire registry.
By bundling several aspects of those levels into a compliance level it could alsoo discard bits of information an attacker can use. Instead of learning that the package maintainers use MFA specifically they will only learn that the maintainers do a whole bunch of things to reach said compliance level.

So here's a rephrased proposal:

```suggestion
  - The package repository requires MFA for all maintainers.
  - The package repository publishes "security profiles" about individual packages
  - The package repository allows individual packages to require MFA for all maintainers. Enabling it is one of the perquisites of marking an individual package as "level 3" in its security profile.
```